File: design-add-and-search-words-data-structure/njngwn.java
class WordDictionary {
    private static class CharDictionary {
        HashMap<Character, CharDictionary> charMap;
        boolean isEnd;

        private CharDictionary() {
            this.charMap = new HashMap<>();
            this.isEnd = false;
        }
    }

    private CharDictionary rootNode;

    public WordDictionary() {
        this.rootNode = new CharDictionary();
    }

    public void addWord(String word) {
        CharDictionary currentNode = this.rootNode;

        for (char ch : word.toCharArray()) {
            if (!currentNode.charMap.containsKey(ch)) {
                currentNode.charMap.put(ch, new CharDictionary());
            }
            currentNode = currentNode.charMap.get(ch);
        }
        currentNode.isEnd = true;
    }

    public boolean search(String word) {
        return searchRecursive(word, this.rootNode, 0);
    }

    private boolean searchRecursive(String word, CharDictionary node, int index) {
        // Base case
        if (index == word.length()) {
            return node.isEnd;
        }

        char ch = word.charAt(index);

        if (ch == '.') {
            for (CharDictionary childNode : node.charMap.values()) {
                if (searchRecursive(word, childNode, index + 1)) {
                    return true;
                }
            }
            return false;
        } else {
            if (!node.charMap.containsKey(ch)) {
                return false;
            }
            return searchRecursive(word, node.charMap.get(ch), index + 1);
        }
    }
}

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary obj = new WordDictionary();
 * obj.addWord(word);
 * boolean param_2 = obj.search(word);
 */

File: design-add-and-search-words-data-structure/njngwn.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.isEnd = False


class WordDictionary:

    def __init__(self):
        self.root = TrieNode()

    def addWord(self, word: str) -> None:
        node = self.root
        for ch in word:
            if ch not in node.children:
                node.children[ch] = TrieNode()
            node = node.children[ch]

        node.isEnd = True

    def search(self, word: str) -> bool:
        def dfs(node: TrieNode, idx: int) -> bool:
            if idx == len(word):
                return node.isEnd

            ch = word[idx]
            if ch == '.':
                for child in node.children.values():
                    if dfs(child, idx + 1):
                        return True
                return False
            else:
                if ch not in node.children:
                    return False
                return dfs(node.children[ch], idx + 1)

        return dfs(self.root, 0)

# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)
